pkg/requesthandler/accounts.go

```diff
@@ -32,26 +35,60 @@ func (r *RequestHandler) CongestionByAccountAddress(accountAddress *iotago.Accou
 	}, nil
 }
 
-func (r *RequestHandler) Validators(slotRange, cursorIndex, pageSize uint32) (*api.ValidatorsResponse, error) {
-	latestCommittedSlot := r.protocol.Engines.Main.Get().SyncManager.LatestCommitment().Slot()
-	latestEpoch := r.protocol.APIForSlot(latestCommittedSlot).TimeProvider().EpochFromSlot(latestCommittedSlot)
-
-	// TODO: Move into the api cache package
-	//registeredValidators, exists := r.protocol.Engines.Main.Get().BlockRetainer.RegisteredValidatorsCache(slotRange)
-	//if !exists {
-	//	registeredValidators, err := r.protocol.Engines.Main.Get().SybilProtection.OrderedRegisteredCandidateValidatorsList(latestEpoch)
-	//	if err != nil {
-	//		return nil, ierrors.Wrapf(echo.ErrInternalServerError, "failed to get ordered registered validators list for epoch %d : %s", latestEpoch, err)
-	//	}
-	//	r.protocol.Engines.Main.Get().BlockRetainer.RetainRegisteredValidatorsCache(slotRange, registeredValidators)
-	//}
-
-	registeredValidators, err := r.protocol.Engines.Main.Get().SybilProtection.OrderedRegisteredCandidateValidatorsList(latestEpoch)
+func (r *RequestHandler) registeredValidatorsFromCache(epochIndex iotago.EpochIndex, key []byte) ([]*api.ValidatorResponse, error) {
+	apiForEpoch := r.APIProvider().APIForEpoch(epochIndex)
+
+	registeredValidatorsBytes := r.registeredValidatorsCache.Get(key)
+	if registeredValidatorsBytes == nil {
+		// get the ordered registered validators list from engine.
+		registeredValidators, err := r.protocol.Engines.Main.Get().SybilProtection.OrderedRegisteredCandidateValidatorsList(epochIndex)
+		if err != nil {
+			return nil, ierrors.Wrapf(echo.ErrNotFound, " ordered registered validators list for epoch %d not found: %s", epochIndex, err)
+		}
+
+		// store validator responses in cache.
+		registeredValidatorsBytes, err := apiForEpoch.Encode(registeredValidators, serix.WithTypeSettings(validatorResponsesTypeSettings))
+		if err != nil {
+			return nil, ierrors.Wrapf(echo.ErrInternalServerError, "failed to encode ordered registered validators list for epoch %d : %s", epochIndex, err)
+		}
+		r.registeredValidatorsCache.Set(key, registeredValidatorsBytes)
+
+		return registeredValidators, nil
+	}
+
+	validatorResp := make([]*api.ValidatorResponse, 0)
```

Comment:
Following the above suggestion we could do decoding in `Cache` logic and already get `[]*api.ValidatorResponse`